### DISCUSSION TITLE
Alarm Status Bounces to "Unavailable" Frequently

### DISCUSSION THREAD

**Discussion by theturtle32** [tsightler/ring-mqtt#336] [Support Q&A]
After upgrading from version 4.6.3 of ring-mqtt to the current version ("latest" tag on docker hub), the alarm status is very frequently showing as "Unavailable" briefly in Home Assistant. I took a screenshot so you can see what I mean.

Is this a known bug? Perhaps related to the upgrade process? What troubleshooting steps would you all recommend?

<img width="1654" alt="Screen Shot 2022-03-17 at 10 21 47 AM" src="https://user-images.githubusercontent.com/53184/158860186-b8d47152-9e47-4b40-bf01-14f5b344e132.png">

**Comment by tsightler** [comment]
There are no known issues of this type and nothing in the code around this has changed for pretty much the entirety of the 4.x releases so it's hard to see how it could be related to the upgrade.  I suppose it could be crashing and restarting, but I don't see the restart pattern in that.   Logs should provide some clue as to what is happening since all online/offline events are logged.

**Reply by theturtle32** [reply]
Looking at the ring-mqtt logs, I do see a fairly regular pattern of these messages sprinkled throughout:

```
ring-mqtt        | 2022-03-18T18:49:52.621Z ring-mqtt Unable to connect to MQTT broker. read ECONNRESET
ring-mqtt        | 2022-03-18T18:49:53.622Z ring-mqtt Attempting to reconnect to MQTT broker...
```

So I looked in the logs for mosquitto (I'm using the `mosquitto:2.0` docker tag) and saw this, which seems to correspond:
```
1647630143: Client mqttjs_71190aed has exceeded timeout, disconnecting.
1647630145: New connection from 172.20.0.1:58040 on port 1883.
1647630145: New client connected from 172.20.0.1:58040 as mqttjs_71190aed (p2, c1, k60, u'ringmqtt').
1647630323: Client mqttjs_71190aed has exceeded timeout, disconnecting.
1647630325: New connection from 172.20.0.1:58354 on port 1883.
1647630325: New client connected from 172.20.0.1:58354 as mqttjs_71190aed (p2, c1, k60, u'ringmqtt').
1647630551: Client mqttjs_71190aed has exceeded timeout, disconnecting.
1647630584: New connection from 172.20.0.1:58828 on port 1883.
1647630584: New client connected from 172.20.0.1:58828 as mqttjs_71190aed (p2, c1, k60, u'ringmqtt').
1647630689: Client mqttjs_71190aed has exceeded timeout, disconnecting.
1647630697: New connection from 172.20.0.1:59042 on port 1883.
1647630697: New client connected from 172.20.0.1:59042 as mqttjs_71190aed (p2, c1, k60, u'ringmqtt').
```

So I guess I'll look at the mosquitto "timeout" configuration options.

**Reply by theturtle32** [reply]
So it turns out there are no configuration options for setting the actual timeout in Mosquitto relating to this error message. So I'm not sure where to go from here.

**Reply by tsightler** [reply]
Pretty strange as the MQTTjs client used by ring-mqtt uses 60 second keepalive by default (I think that's what the 'k60' value is indicating even), which should be more than enough to keep mosquitto from kicking out clients (I believe default is 90 seconds).  Clearly something is going on here from a network perspective, but not sure what that could be as this is very basic stuff.

**Reply by tsightler** [reply]
Or I guess it could be a bug in MQTT.js library, although I've had no other reports of this behavior and I've not seen it at all, however, it does look quite a bit like this issue:

https://github.com/mqttjs/MQTT.js/issues/1362

**Reply by tsightler** [reply]
You can enable debug level logging on the mosquitto side and on the ring-mqtt side.  On ring-mqtt set DEBUG=mqttjs* to see debug output from the MQTTjs library.  You should see the keepalive pings (PINGREQ/PINGRESP) and, if you do the same on the mosquitto server, it might provide some additional insight.

Another thing to check, are the two systems running on the same server?  Timesync can be weirdly important in this case.  Longshot I know, but I thought I would mention it.

**Reply by theturtle32** [reply]
Yes, they're running on the same server in different containers.

I've resolved the issue, but I'm still not sure what it was. I simplified/revamped the mosquitto container definition in my docker-compose.yml and it started working reliably, so I went through and put it back to the way it was, one line at a time, to try to isolate which change was the problem, and ended up getting the configuration all the way back to the way it was originally without ever reproducing the timeout disconnection issue again.

So, I'm kind of at a loss. I guess I'll chalk it up to some ephemeral Docker networking issue?

**Reply by theturtle32** [reply]
In case you're curious, this is the current configuration -- several of these lines are unnecessary, and I'll remove them going forward, but wanted to show it as-is. (The config was originally generated by [docker-autocompose](https://github.com/Red5d/docker-autocompose))

```yaml
  mosquitto:
     container_name: mosquitto
     image: eclipse-mosquitto:2.0
     logging:
       driver: json-file
       options:
         max-size: 10m
         max-file: "3"
     tty: true
     stdin_open: true
     mac_address: 02:42:ac:11:00:02
     ports:
       - 1883:1883/tcp
       - 9001:9001/tcp
     volumes:
       - /home/ubuntu/mosquitto/data:/mosquitto/data
       - /home/ubuntu/mosquitto/log:/mosquitto/log
       - /home/ubuntu/mosquitto/config:/mosquitto/config:ro
     restart: unless-stopped
```

**Reply by tsightler** [reply]
Thanks for the update, glad you got it working!
